File: clients/go/orm/aes_rotation.go
package orm

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

// AESKeyring stores the keys that can decode persisted AES values. Versions
// are explicit so a rotation can read one row with its stored version and
// write every AES column with one target version.
type AESKeyring struct {
	keys    map[int32]string
	current int32
}

// aesKeyring builds the keyring of the connection configuration once.
func (d *DB) aesKeyring() (AESKeyring, error) {
	d.m.keyringOnce.Do(func() {
		keys := d.cfg.AESKeys
		if len(keys) == 0 && d.cfg.AESKey != "" {
			keys = map[int32]string{d.cfg.AESVersion: d.cfg.AESKey}
		}
		d.m.keyring, d.m.keyringErr = NewAESKeyring(keys, d.cfg.AESVersion)
	})
	return d.m.keyring, d.m.keyringErr
}

// HostDecodeVersioned decodes an AES value with the row's stored version.
func HostDecodeVersioned(raw any, styles []string, version int32, keyring AESKeyring) (any, error) {
	key, ok := keyring.keys[version]
	if !ok {
		return nil, fmt.Errorf("aes key version %d is not declared", version)
	}
	return hostDecode(raw, styles, key)
}

// NewAESKeyring validates a non-empty key for every declared version and a
// current version that is present in the key list.
func NewAESKeyring(keys map[int32]string, current int32) (AESKeyring, error) {
	copyKeys := make(map[int32]string, len(keys))
	for version, key := range keys {
		if version < 1 || key == "" {
			return AESKeyring{}, fmt.Errorf("aes key version %d has an empty key or is invalid", version)
		}
		copyKeys[version] = key
	}
	if current < 1 || copyKeys[current] == "" {
		return AESKeyring{}, fmt.Errorf("aes current version %d is not declared", current)
	}
	return AESKeyring{keys: copyKeys, current: current}, nil
}

// CurrentAESVersion returns the target version used by new writes.
func (k AESKeyring) CurrentAESVersion() int32 { return k.current }

// AESVersions returns declared versions in deterministic order.
func (k AESKeyring) AESVersions() []int32 {
	out := make([]int32, 0, len(k.keys))
	for version := range k.keys {
		out = append(out, version)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out
}

// RotateAESRow returns a copy with every AES column re-encrypted. The input
// row is unchanged until all columns decode and encode successfully. A caller
// must persist the returned columns and version in one database transaction.
func RotateAESRow(row map[string]any, versionColumn string, columns []AESRotationColumn, targetVersion int32, keyring AESKeyring) (map[string]any, error) {
	oldVersion, err := rowVersion(row[versionColumn])
	if err != nil {
		return nil, err
	}
	oldKey, ok := keyring.keys[oldVersion]
	if !ok {
		return nil, fmt.Errorf("aes key version %d is not declared", oldVersion)
	}
	newKey, ok := keyring.keys[targetVersion]
	if !ok {
		return nil, fmt.Errorf("aes target version %d is not declared", targetVersion)
	}
	out := make(map[string]any, len(row)+1)
	for name, value := range row {
		out[name] = value
	}
	for _, column := range columns {
		value, exists := row[column.Name]
		if !exists {
			return nil, fmt.Errorf("aes column %s is missing", column.Name)
		}
		plain, err := hostDecode(value, column.Styles, oldKey)
		if err != nil {
			return nil, fmt.Errorf("aes column %s: %w", column.Name, err)
		}
		encoded, err := HostEncode(plain, column.Styles, newKey)
		if err != nil {
			return nil, fmt.Errorf("aes column %s: %w", column.Name, err)
		}
		out[column.Name] = encoded
	}
	out[versionColumn] = targetVersion
	return out, nil
}

func quoteIdentifier(driver, value string) string {
	if value == "" || strings.ContainsAny(value, "\x00\"`.;'") {
		panic("invalid generated SQL identifier")
	}
	if driver == "mysql" {
		return "`" + value + "`"
	}
	return "\"" + value + "\""
}

func rowVersion(value any) (int32, error) {
	var version int64
	switch v := value.(type) {
	case int:
		version = int64(v)
	case int8:
		version = int64(v)
	case int16:
		version = int64(v)
	case int32:
		version = int64(v)
	case int64:
		version = v
	case uint:
		if uint64(v) > math.MaxInt32 {
			return 0, fmt.Errorf("aes version %d is out of range", v)
		}
		version = int64(v)
	case uint8:
		version = int64(v)
	case uint16:
		version = int64(v)
	case uint32:
		version = int64(v)
	case uint64:
		if v > math.MaxInt32 {
			return 0, fmt.Errorf("aes version %d is out of range", v)
		}
		version = int64(v)
	case string:
		n, err := strconv.ParseInt(v, 10, 64)
		if err != nil {
			return 0, fmt.Errorf("aes version %q is not an integer", v)
		}
		version = n
	default:
		return 0, fmt.Errorf("aes version has type %T", value)
	}
	if version < 1 || version > math.MaxInt32 {
		return 0, fmt.Errorf("aes version %d is out of range", version)
	}
	return int32(version), nil
}

File: clients/rust/orm/src/aes_rotation.rs
use std::collections::BTreeMap;

use crate::db::Exec;
use crate::plan::{BindSlot, Step};
use crate::row::read_row;
use crate::value::Param;
use crate::value::Val;
use crate::{Error, Result};

#[derive(Debug, Clone)]
pub struct AesRotationColumn {
    pub name: String,
    pub styles: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct AesRotationSpec {
    pub table: String,
    pub primary_keys: Vec<String>,
    pub version_column: String,
    pub columns: Vec<AesRotationColumn>,
    pub batch_size: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AesRotationStatus {
    pub current: i32,
    pub total: i64,
    pub pending: i64,
    pub versions: BTreeMap<i32, i64>,
}

#[derive(Debug, Clone)]
pub struct AesKeyring {
    keys: BTreeMap<i32, String>,
    pub current_version: i32,
}

impl AesKeyring {
    pub fn new(keys: BTreeMap<i32, String>, current_version: i32) -> Result<Self> {
        if keys
            .iter()
            .any(|(version, key)| *version < 1 || key.is_empty())
        {
            return Err(Error::Config(
                "AES key list contains an invalid entry".into(),
            ));
        }
        if !keys.contains_key(&current_version) {
            return Err(Error::Config(format!(
                "AES version {current_version} is not declared"
            )));
        }
        Ok(Self {
            keys,
            current_version,
        })
    }

    pub fn versions(&self) -> Vec<i32> {
        self.keys.keys().copied().collect()
    }

    pub fn key(&self, version: i32) -> Result<&str> {
        self.keys
            .get(&version)
            .map(String::as_str)
            .ok_or_else(|| Error::Config(format!("AES version {version} is not declared")))
    }

    /// Returns a copy after every AES column succeeds. The caller must persist
    /// all returned columns and the version in one database transaction.
    pub fn rotate_row<D, E>(
        &self,
        row: &BTreeMap<String, Val>,
        version_column: &str,
        columns: &[AesRotationColumn],
        target_version: i32,
        decode: D,
        encode: E,
    ) -> Result<BTreeMap<String, Val>>
    where
        D: Fn(&Val, &[String], &str) -> Result<Val>,
        E: Fn(&Val, &[String], &str) -> Result<Val>,
    {
        let old_version = match row.get(version_column) {
            Some(Val::I64(value)) if *value >= 1 && *value <= i32::MAX as i64 => *value as i32,
            _ => return Err(Error::Config("AES row version must be an integer".into())),
        };
        let old_key = self
            .keys
            .get(&old_version)
            .ok_or_else(|| Error::Config(format!("AES version {old_version} is not declared")))?;
        let new_key = self.keys.get(&target_version).ok_or_else(|| {
            Error::Config(format!("AES version {target_version} is not declared"))
        })?;
        let mut out = row.clone();
        for column in columns {
            let value = row
                .get(&column.name)
                .ok_or_else(|| Error::Config(format!("AES column {} is missing", column.name)))?;
            let plain = decode(value, &column.styles, old_key)?;
            out.insert(
                column.name.clone(),
                encode(&plain, &column.styles, new_key)?,
            );
        }
        out.insert(version_column.to_owned(), Val::I64(target_version as i64));
        Ok(out)
    }
}

fn quote(driver: &str, value: &str) -> Result<String> {
    if value.is_empty()
        || !value
            .bytes()
            .all(|byte| byte == b'_' || byte.is_ascii_alphanumeric())
        || value.as_bytes()[0].is_ascii_digit()
    {
        return Err(Error::Config(format!(
            "invalid generated identifier {value}"
        )));
    }
    Ok(if driver == "mysql" {
        format!("`{value}`")
    } else {
        format!("\"{value}\"")
    })
}

fn placeholder(driver: &str, position: usize) -> String {
    if driver == "postgres" {
        format!("${position}")
    } else {
        "?".into()
    }
}
fn step(sql: String, binds: Vec<BindSlot>) -> Step {
    Step {
        plan_id: 0,
        id: 0,
        role: "aes_rotation".into(),
        sql,
        lock: String::new(),
        bind_slots: binds,
        assemble: None,
        parent: None,
    }
}
fn parameter(position: usize) -> BindSlot {
    BindSlot {
        from: "param".into(),
        param: position,
        transform: String::new(),
        name: String::new(),
        step: 0,
        column: String::new(),
        host_styles: vec![],
        col_type: String::new(),
    }
}
fn param(value: &Val) -> Result<Param> {
    Ok(match value {
        Val::Null => Param::Null,
        Val::I64(v) => Param::I64(*v),
        Val::F64(v) => Param::F64(*v),
        Val::Str(v) => Param::Str(v.clone()),
        Val::Bytes(v) => Param::Bytes(v.clone()),
        Val::DateTime(v) => Param::DateTime(*v),
        Val::Date(v) => Param::Date(*v),
        Val::Bool(v) => Param::Bool(*v),
        Val::Json(v) => Param::Str(v.to_string()),
    })
}
fn val(value: Param) -> Val {
    match value {
        Param::Null => Val::Null,
        Param::I64(v) => Val::I64(v),
        Param::F64(v) => Val::F64(v),
        Param::Str(v) => Val::Str(v),
        Param::Bytes(v) => Val::Bytes(v),
        Param::DateTime(v) => Val::DateTime(v),
        Param::Date(v) => Val::Date(v),
        Param::Bool(v) => Val::Bool(v),
        Param::Point(v) => Val::Str(format!("POINT({} {})", v.0, v.1)),
    }
}
fn encode(value: &Val, styles: &[String], key: &str) -> Result<Val> {
    Ok(val(crate::codec::host_encode(&param(value)?, styles, key)?))
}

pub async fn aes_status(
    ex: &impl Exec,
    spec: &AesRotationSpec,
    keyring: &AesKeyring,
) -> Result<AesRotationStatus> {
    let driver = ex.db().driver();
    let table = quote(driver, &spec.table)?;
    let version = quote(driver, &spec.version_column)?;
    let rows = ex
        .query(
            &step(
                format!(
                    "SELECT {version}, COUNT(*) FROM {table} GROUP BY {version} ORDER BY {version}"
                ),
                vec![],
            ),
            &[],
            vec![],
        )
        .await?;
    let mut status = AesRotationStatus {
        current: keyring.current_version,
        total: 0,
        pending: 0,
        versions: BTreeMap::new(),
    };
    for row in rows {
        let values = read_row(&row, 2)?;
        let stored_value = values[0].as_i64();
        let count = values[1].as_i64();
        if !(1..=i32::MAX as i64).contains(&stored_value) || count < 0 {
            return Err(Error::Engine {
                code: crate::codes::CODEC_DECODE.into(),
                msg: "AES rotation status contains an invalid value".into(),
            });
        }
        let stored = stored_value as i32;
        status.versions.insert(stored, count);
        status.total += count;
        if stored != status.current {
            status.pending += count;
        }
    }
    Ok(status)
}

pub async fn rotate_aes_rows(
    ex: &impl Exec,
    spec: &AesRotationSpec,
    keyring: &AesKeyring,
) -> Result<u64> {
    if ex.tx().is_none() {
        let spec = spec.clone();
        let keyring = keyring.clone();
        return ex
            .db()
            .transaction(move |tx| {
                let spec = spec.clone();
                let keyring = keyring.clone();
                async move { rotate_in_transaction(&tx, &spec, &keyring).await }
            })
            .await;
    }
    rotate_in_transaction(ex, spec, keyring).await
}

async fn rotate_in_transaction(
    ex: &impl Exec,
    spec: &AesRotationSpec,
    keyring: &AesKeyring,
) -> Result<u64> {
    if spec.columns.is_empty() {
        return Err(Error::Config("AES rotation columns are empty".into()));
    }
    if spec.primary_keys.is_empty() {
        return Err(Error::Config("AES rotation primary keys are empty".into()));
    }
    let driver = ex.db().driver();
    let table = quote(driver, &spec.table)?;
    let primary: Vec<String> = spec
        .primary_keys
        .iter()
        .map(|key| quote(driver, key))
        .collect::<Result<_>>()?;
    let version = quote(driver, &spec.version_column)?;
    let columns: Vec<String> = spec
        .columns
        .iter()
        .map(|column| quote(driver, &column.name))
        .collect::<Result<_>>()?;
    let batch_size = if spec.batch_size == 0 {
        1000
    } else {
        spec.batch_size
    };
    let select = step(format!("SELECT {}, {version}, {} FROM {table} WHERE {version} <> {} ORDER BY {} LIMIT {batch_size}", primary.join(", "), columns.join(", "), placeholder(driver, 1), primary.join(", ")), vec![parameter(0)]);
    let rows = ex
        .query(
            &select,
            &[Param::I64(keyring.current_version as i64)],
            vec![],
        )
        .await?;
    let mut sets: Vec<String> = columns
        .iter()
        .enumerate()
        .map(|(index, name)| format!("{name} = {}", placeholder(driver, index + 1)))
        .collect();
    sets.push(format!(
        "{version} = {}",
        placeholder(driver, columns.len() + 1)
    ));
    let mut predicates: Vec<String> = primary
        .iter()
        .enumerate()
        .map(|(index, key)| format!("{key} = {}", placeholder(driver, columns.len() + 2 + index)))
        .collect();
    predicates.push(format!(
        "{version} = {}",
        placeholder(driver, columns.len() + 2 + primary.len())
    ));
    let bind_count = columns.len() + primary.len() + 2;
    let update = step(
        format!(
            "UPDATE {table} SET {} WHERE {}",
            sets.join(", "),
            predicates.join(" AND ")
        ),
        (0..bind_count).map(parameter).collect(),
    );
    let mut count = 0;
    for row in rows {
        let values = read_row(&row, primary.len() + columns.len() + 1)?;
        let mut source = BTreeMap::new();
        for (index, key) in spec.primary_keys.iter().enumerate() {
            source.insert(key.clone(), values[index].clone());
        }
        source.insert(spec.version_column.clone(), values[primary.len()].clone());
        for (index, column) in spec.columns.iter().enumerate() {
            source.insert(
                column.name.clone(),
                values[index + primary.len() + 1].clone(),
            );
        }
        let rotated = keyring.rotate_row(
            &source,
            &spec.version_column,
            &spec.columns,
            keyring.current_version,
            crate::codec::host_decode,
            encode,
        )?;
        let mut args = Vec::with_capacity(bind_count);
        for column in &spec.columns {
            args.push(param(&rotated[&column.name])?);
        }
        args.push(Param::I64(keyring.current_version as i64));
        for value in &values[..primary.len()] {
            args.push(param(value)?);
        }
        args.push(Param::I64(values[primary.len()].as_i64()));
        let (_, affected) = ex.execute(&update, &args).await?;
        if affected != 1 {
            return Err(Error::Engine {
                code: crate::codes::OPTIMISTIC_LOCK.into(),
                msg: format!(
                    "AES rotation changed {} primary key {:?}",
                    spec.table,
                    &values[..primary.len()]
                ),
            });
        }
        count += 1;
    }
    Ok(count)
}
